apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/LruTypePoolCache.java

```diff
@@ -0,0 +1,279 @@
+            .scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    evictStaleEntries(maxAgeMs);
+                }
+            }, maxAgeMs, maxAgeMs, TimeUnit.MILLISECONDS);
+        return this;
+    }
+
+    void evictStaleEntries(long maxAgeMs) {
+        long deadline = System.currentTimeMillis() - maxAgeMs;
+        for (Iterator<Map.Entry<String, ResolutionsByClassLoader>> iterator = getSharedCache().entrySet().iterator(); iterator.hasNext(); ) {
+            Map.Entry<String, ResolutionsByClassLoader> entry = iterator.next();
+            if (entry.getValue().isExpired(deadline)) {
+                iterator.remove();
+            }
+        }
+        cacheProviders.expungeStaleEntries();
+    }
+
+    private ConcurrentMap<String, ResolutionsByClassLoader> createCache() {
+        return new ConcurrentLinkedHashMap.Builder<String, ResolutionsByClassLoader>()
+            .maximumWeightedCapacity(maxCacheSize)
+            .weigher(new EntryWeigher<String, ResolutionsByClassLoader>() {
+                @Override
+                public int weightOf(String key, ResolutionsByClassLoader value) {
+                    return Math.max(1, value.size());
+                }
+            })
+            .build();
+    }
+
+    @Override
+    protected TypePool.CacheProvider locate(@Nullable ClassLoader classLoader) {
+        if (classLoader == null) {
+            classLoader = getBootstrapMarkerLoader();
+        }
+        TypePool.CacheProvider cacheProvider = cacheProviders.get(classLoader);
+        if (cacheProvider == null) {
+            cacheProvider = new GlobalCacheProviderAdapter(new WeakReference<>(classLoader), this);
+            TypePool.CacheProvider racy = cacheProviders.putIfAbsent(classLoader, cacheProvider);
+            if (racy != null) {
+                cacheProvider = racy;
+            }
+        }
+        return cacheProvider;
+    }
+
+    /**
+     * Copied from {@code net.bytebuddy.agent.builder.AgentBuilder.PoolStrategy.WithTypePoolCache.Simple#getBootstrapMarkerLoader()}
+     * <p>
+     * Returns the class loader to serve as a cache key if a cache provider for the bootstrap class loader is requested.
+     * This class loader is represented by {@code null} in the JVM which is an invalid value for many {@link ConcurrentMap}
+     * implementations.
+     * </p>
+     * <p>
+     * By default, {@link ClassLoader#getSystemClassLoader()} is used as such a key as any resource location for the
+     * bootstrap class loader is performed via the system class loader within Byte Buddy as {@code null} cannot be queried
+     * for resources via method calls such that this does not make a difference.
+     * </p>
+     *
+     * @return A class loader to represent the bootstrap class loader.
+     */
+    private ClassLoader getBootstrapMarkerLoader() {
+        return ClassLoader.getSystemClassLoader();
```

Comment:
[question] If I understand it correctly, it's only safe/correct to use the system CL here because we are just using it to load resources, and not to load classes themselves. Thus, it's a bit more than just a "marker" for map key.

Reply:
I'd like to stick with the terminology used within Byte Buddy itself, which is where I copied this method from.